File: svn/branches/game/CaseyPyroJ/src/edu/wpi/first/wpilibj/templates/Driver.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */

package edu.wpi.first.wpilibj.templates;

import edu.wpi.first.wpilibj.*;

/**
 *
 * @author Currey-WilsonC
 */
public class Driver implements Part {

    //mecanum tuning constants
    private double frP = 0.976161;
    private double frN = 0.976641;
    private double flP = 0.927171;
    private double flN = 0.888988;
    private double brP = 1.000000;
    private double brN = 0.980499;
    private double blP = 0.997739;
    private double blN = 0.967600;
    
    private double x;
    private double y;
    private double r;
    
    public boolean enableGyroCorrect;

    private double flstart;
    private double frstart;
    private double blstart;
    private double brstart;

    private double frdelta;
    private double fldelta;
    private double brdelta;
    private double bldelta;

    private double negfrdelta;
    private double negfldelta;
    private double negbrdelta;
    private double negbldelta;

    int tmr;
    
    public Driver()
    {
        x = y = r = 0;
        tmr = 0;
    }

    public void runDiagnostics(PyroMain main){}

    public void drive(double x, double y, double r)
    {
        this.x = x;
        this.y = y;
        this.r = r;
    }

    public void setX(double x) { this.x = x; }
    public void setY(double y) { this.y = y; }
    public void setR(double r) { this.r = r; }
    public double getX() { return x; }
    public double getY() { return y; }
    public double getR() { return r; }

    public void initAutonomous(PyroMain main)
    {
        x = y = r = 0;
    }
    public void initOperator(PyroMain main)
    {
        x = y = r = 0;
    }
    public void runAutonomous(PyroMain main)
    {
        updateDriver(main);
    }
    public void runOperator(PyroMain main)
    {
        updateDriver(main);
    }

    private void SETALL(double pwr) { drive(0,pwr,0); }
    private double SQUARE(double X) { return (X*X); }
    private double MINIMUM(double X,double Y) { return Math.min(X,Y); }
    private double GETTUNECONST(double m,double var) { return (((double)m) / ((double)var)); }
    /*private void GETEND(PyroMain main)
    {
        frdelta = Math.abs(main.hal.getFREncoder() - frStart );
        fldelta = Math.abs(main.hal.getFLEncoder() - flStart );
        brdelta = Math.abs(main.hal.getBREncoder() - brStart );
        bldelta = Math.abs(main.hal.getBLEncoder() - blStart );
    }
/*
    public void updateTune(PyroMain main)
    {
        //enum tuneMode { tuneIdle, tuneRev, tuneForward, tunePause, tuneRevBackward, tuneBackward, tuneStop };
        int mode = 0;
	Timer timer = new Timer();
	//static int frStart, flStart, brStart, blStart;
	double tunePwr = 1.0;
	double revTime = 1.5;
	double driveTime = 2.5;
	if (main.hal.lStick.getTop() && main.hal.rStick.getTop()) {
		//System.out.println("mec tune");
		switch (mode) {
		case 0:
			System.out.println("starting idle!\n");
			timer.start();
			mode = 1;
			break;
		case 1:
			SETALL(tunePwr);
			if (timer.get() > revTime) {
				timer.reset();
				mode = 2;
				//getENC
				main.hal.resetEncoders();
				System.out.println ("Got rev encoders:\n");
				System.out.println ("FR: %i\n", frStart);
				System.out.println ("FL: %i\n", flStart);
				System.out.println ("BR: %i\n", brStart);
				System.out.println ("BL: %i\n", blStart);
				System.out.println("Done reving up\n");
			}
			break;
		case 2:
			SETALL(tunePwr);
			if (timer.get() > driveTime) {
				timer.reset();
				mode = 3;
				GETEND(main);
				System.out.println ("Forward:\n");
                                //format is fr,fl,br,bl
				System.out.println(frdelta);
				System.out.println (fldelta);
				System.out.println (brdelta);
				System.out.println (bldelta);
				System.out.println("\n");
			}
			break;
		case 3:
			SETALL(0.0);
			if (timer.get() > 1.5) {
				timer.reset();
				mode = 4;
			}
			break;
		case 4:
			SETALL(-tunePwr);
			if (timer.get() > revTime) {
				timer.reset();
				mode = 5;
				//getENC
				main.hal.resetEncoders();
				System.out.println("Done reving up\n");
			}
			break;
		case 5:
			SETALL(-tunePwr);
			if (timer.get() > driveTime) {
				timer.reset();
				mode = 6;
				GETEND(main);
				double m = (double)MINIMUM(MINIMUM(MINIMUM(frdelta,negfrdelta),MINIMUM(fldelta,negfldelta)),MINIMUM(MINIMUM(brdelta,negbrdelta),MINIMUM(bldelta,negbldelta)));
				System.out.println ("Backward:\n");
				System.out.println (negfrdelta);
				System.out.println (negfldelta);
				System.out.println (negbrdelta);
				System.out.println (negbldelta);
				System.out.println("\n");
				System.out.println("=================\n");
				System.out.println("TUNING CONSTANTS:\n");
				System.out.println("=================\n");
				System.out.println(GETTUNECONST(frdelta,m));
				System.out.println(GETTUNECONST(negfrdelta,m));
				System.out.println(GETTUNECONST(fldelta,m));
				System.out.println(GETTUNECONST(negfldelta,m));
				System.out.println(GETTUNECONST(brdelta,m));
				System.out.println(GETTUNECONST(negbrdelta,m));
				System.out.println(GETTUNECONST(bldelta,m));
				System.out.println(GETTUNECONST(negbldelta,m));
				double avg = (((double)frdelta)+
						     ((double)negfrdelta)+
						     ((double)fldelta)+
						     ((double)negfldelta)+
						     ((double)brdelta)+
						     ((double)negbrdelta)+
						     ((double)bldelta)+
						     ((double)negbldelta)) / 8.0;
				// standard deviation-ish thingy
				double deviance = SQUARE(frdelta-avg)+
					           SQUARE(negfrdelta-avg)+
					           SQUARE(fldelta-avg)+
					           SQUARE(negfldelta-avg)+
					           SQUARE(brdelta-avg)+
					           SQUARE(negbrdelta-avg)+
					           SQUARE(bldelta-avg)+
					           SQUARE(negbldelta-avg);
                                System.out.println("Deviance:");
				System.out.println(deviance);
			}
			break;
		case 6:
			SETALL(0.0);
		}
	} else {
		timer.reset();
		timer.stop();
		mode = 0;
	}
    }
*/
    public void updateDriver(PyroMain main)
    {
        //double beginStrafeAngle;
	//boolean wasStrafing = false;

	//updateTune(main);

	boolean isStrafing = r == 0 && (y != 0 || x != 0);
	//double strafeAngle = main.hal.getGyroValue();

	//if (isStrafing && !wasStrafing)
            //beginStrafeAngle = strafeAngle;

        double fl = y + x + r;
	double fr = y - x - r;
	double bl = y - x + r;
	double br = y + x - r;

        /*#define TUNE_MTR(MTR) if (MTR > 0) MTR *= MTR ## P; else if (MTR < 0) MTR *= MTR ## N;
		TUNE_MTR(fr)
		TUNE_MTR(fl)
		TUNE_MTR(br)
		TUNE_MTR(bl)*/
        if (fr > 0)
            fr *= frP;
        else
            fr *= frN;
        if (fl > 0)
            fl *= flP;
        else
            fl *= flN;
        if (br > 0)
            br *= brP;
        else
            br *= brN;
        if (bl > 0)
            bl *= blP;
        else
            bl *= blN;

	if (tmr++ > 5) {
            tmr=0;
            //printf("%d %d %d %d\n",fr,fl,br,bl);
	}

        main.hal.setFLPower(fl);
        main.hal.setFRPower(fr);
        main.hal.setBLPower(bl);
        main.hal.setBRPower(br);

        //wasStrafing = isStrafing;
    }
}

File: svn/tags/game/Spy-Pyro/Driver.cpp
#include "Driver.h"
#include "WPILib.h"
#include "HAL.h"

Driver* const driver = new Driver();

void Driver::updateTune()
{
#define SETALL(pwr) drive(0,pwr,0);
/*#define GETENC \
	frStart = hal->getFREncoder();\
	flStart = hal->getFLEncoder();\
	brStart = hal->getBREncoder();\
	blStart = hal->getBLEncoder();*/
#define GETEND(prefix) \
	    prefix ## frDelta = abs(hal->getFREncoder() /*- frStart*/ ),\
	    prefix ## flDelta = abs(hal->getFLEncoder() /*- flStart*/ ),\
	    prefix ## brDelta = abs(hal->getBREncoder() /*- brStart*/ ),\
	    prefix ## blDelta = abs(hal->getBLEncoder() /*- blStart*/ );
#define MINIMUM(a,b) ((a) > (b) ? (b) : (a))
	enum tuneMode { tuneIdle, tuneRev, tuneForward, tunePause, tuneRevBackward, tuneBackward, tuneStop };
	static tuneMode mode = tuneIdle;
	static Timer timer;
	//static int frStart, flStart, brStart, blStart;
	static int posfrDelta, negfrDelta,
			   posflDelta, negflDelta,
			   posbrDelta, negbrDelta,
			   posblDelta, negblDelta;
	const float tunePwr = 1.0;
	const float revTime = 1.5;
	const float driveTime = 2.5;
	if (hal->lStick.GetTop() && hal->rStick.GetTop()) {
		//printf("mec tune");
		switch (mode) {
		case tuneIdle:
			printf("starting idle!\n");
			timer.Start();
			mode = tuneRev;
			break;
		case tuneRev:
			SETALL(tunePwr);
			if (timer.Get() > revTime) {
				timer.Reset();
				mode = tuneForward;
				//GETENC
				hal->resetEncoders();
				/*printf ("Got rev encoders:\n");
				printf ("FR: %i\n", frStart);
				printf ("FL: %i\n", flStart);
				printf ("BR: %i\n", brStart);
				printf ("BL: %i\n", blStart);*/
				printf("Done reving up\n");
			}
			break;
		case tuneForward:
			SETALL(tunePwr);
			if (timer.Get() > driveTime) {
				timer.Reset();
				mode = tunePause;
				GETEND(pos)
				printf ("Forward:\n");
				printf ("FR: %i\n", posfrDelta);
				printf ("FL: %i\n", posflDelta);
				printf ("BR: %i\n", posbrDelta);
				printf ("BL: %i\n", posblDelta);
				printf("\n");
			}
			break;
		case tunePause:
			SETALL(0.0);
			if (timer.Get() > 1.5) {
				timer.Reset();
				mode = tuneRevBackward;
			}
			break;
		case tuneRevBackward:
			SETALL(-tunePwr);
			if (timer.Get() > revTime) {
				timer.Reset();
				mode = tuneBackward;
				//GETENC
				hal->resetEncoders();
				printf("Done reving up\n");
			}
			break;
		case tuneBackward:
			SETALL(-tunePwr);
			if (timer.Get() > driveTime) {
				timer.Reset();
				mode = tuneStop;
				GETEND(neg)
				float m = (float) MINIMUM(
						MINIMUM(MINIMUM(posfrDelta,negfrDelta),
								MINIMUM(posflDelta,negflDelta)),
						MINIMUM(MINIMUM(posbrDelta,negbrDelta),
								MINIMUM(posblDelta,negblDelta)));
				printf ("Backward:\n");
				printf ("FR: %i\n", negfrDelta);
				printf ("FL: %i\n", negflDelta);
				printf ("BR: %i\n", negbrDelta);
				printf ("BL: %i\n", negblDelta);
				printf("\n");
				printf("=================\n");
				printf("TUNING CONSTANTS:\n");
				printf("=================\n");
#define GETTUNECONST(var) (((float)m) / ((float)var))
				printf("frPos = %f\n", GETTUNECONST(posfrDelta));
				printf("frNeg = %f\n", GETTUNECONST(negfrDelta));
				printf("flPos = %f\n", GETTUNECONST(posflDelta));
				printf("flNeg = %f\n", GETTUNECONST(negflDelta));
				printf("brPos = %f\n", GETTUNECONST(posbrDelta));
				printf("brNeg = %f\n", GETTUNECONST(negbrDelta));
				printf("blPos = %f\n", GETTUNECONST(posblDelta));
				printf("blNeg = %f\n", GETTUNECONST(negblDelta));
				float avg = (((float)posfrDelta)+
						     ((float)negfrDelta)+
						     ((float)posflDelta)+
						     ((float)negflDelta)+
						     ((float)posbrDelta)+
						     ((float)negbrDelta)+
						     ((float)posblDelta)+
						     ((float)negblDelta)) / 8.0;
#define SQUARE(X) ((X) * (X))
				// standard deviation-ish thingy
				float deviance = SQUARE(posfrDelta-avg)+
					           SQUARE(negfrDelta-avg)+
					           SQUARE(posflDelta-avg)+
					           SQUARE(negflDelta-avg)+
					           SQUARE(posbrDelta-avg)+
					           SQUARE(negbrDelta-avg)+
					           SQUARE(posblDelta-avg)+
					           SQUARE(negblDelta-avg);
				
				printf("deviance = %f\n",deviance);
			}
			break;
		case tuneStop:
			SETALL(0.0);
		}
	} else {
		timer.Reset();
		timer.Stop();
		mode = tuneIdle;
	}
}

void Driver::updateDriver()
{
	static float beginStrafeAngle;
	static bool wasStrafing = false;
	
	updateTune();
	
	bool isStrafing = r == 0 && (y != 0 || x != 0);
	float strafeAngle = hal->getGyroValue();
	
	if (isStrafing && !wasStrafing)
		beginStrafeAngle = strafeAngle;
#if 0
	if (isStrafing && !wasStrafing && enableGyroCorrect) {
		float delta = strafeAngle - beginStrafeAngle;
		const float k = 0.05;
		const float maxCorrect = 0.4;
		float correct = delta * k;
		if (correct > maxCorrect) correct = maxCorrect;
		if (correct < -maxCorrect) correct = -maxCorrect;
		r += correct;
	}
#endif
	float fl = y + x + r;
	float fr = y - x - r;
	float bl = y - x + r;
	float br = y + x - r;
#if 1
	//if (hal->rStick.GetTrigger()) {
		// mecanum tuning
		#define TUNE_MTR(MTR) if (MTR > 0) MTR *= MTR ## P; else if (MTR < 0) MTR *= MTR ## N;
		TUNE_MTR(fr)
		TUNE_MTR(fl)
		TUNE_MTR(br)
		TUNE_MTR(bl)
	//}
#endif
	
    // The most awesome mecanum code, written by Tony and Stanford and Rohan
	static int tmr = 0;
	if (tmr++ > 5) {
		tmr=0;
		//printf("%f %f %f %f\n",fr,fl,br,bl);
	}
    hal->setFLPower(fl);
    hal->setFRPower(fr);
    hal->setBLPower(bl);
    hal->setBRPower(br);
    
    wasStrafing = isStrafing;
}
